napari/conftest.py

```diff
@@ -549,3 +551,23 @@ def pytest_collection_modifyitems(session, config, items):
                 index = i
         test_order[index].append(item)
     items[:] = list(chain(*test_order))
+
+
+@pytest.fixture(autouse=True)
+def disable_notification_dismiss_timer(monkeypatch):
+    """
+    This fixture disables starting timer for closing notification
+    by setting the value of `NapariQtNotification.DISMISS_AFTER` to 0.
+
+    As Qt timer is realised by thread and keep reference to the object,
```

Comment:
From the [description of `QTimer`](https://doc.qt.io/qt-5/qtimer.html#details), it's not clear that it uses a thread to achieve its asynchrony. In fact, the description suggests that it may in fact rely on controlling the idle time of the event loop it is associated with (because it seems using a duration of 0 was a way to achieve time slicing on the main thread while keeping the GUI responsive before multi-threading was a thing). A [related SO post](https://stackoverflow.com/questions/42279360/does-a-qtimer-object-run-in-a-separate-thread-what-is-its-mechanism) suggests the same too.

But if you have some docs or evidence to show that it does use a thread, then I understand the explanation as follows.

1. A test creates an instance of `NapariQtNotification`, which is a Python object ultimately backed by a C++ instance of `QDialog`. Refer to this as `self`.
2. We create an instance of `QTimer`, which is a Python object backed by a C++ instance of `QTimer`, and store it as `self.timer`.
3. We connect `self.timer.timeout` to `self.close`, which does not increase the reference count of `self` because  all the logic is resolved in C++ code, which is unaware that `self` is a Python object. 
4. The test finishes and the reference count to `self` drops to 0, and is then garbage collected some time after, which also deletes the C++ instance of `QDialog`.
5. The timeout signal is emitted by the `QTimer` on the non-main thread, and the connection either tries to access the C++ instance of `QDialog` (which has been deleted) when working out on which thread to execute the slot or tries to access the Python instance of `NapariQtNotification` (which is also gone) when executing the slot.

The [`QObject` destructor](https://doc.qt.io/qt-5/qobject.html#dtor.QObject) should disconnect all signals, but I guess that the signal could be emitted before that, and the slot gets executed on the main thread after the destructor has finished? I'd hope/expect Qt to handle that situation safely, but I might be missing something or overestimating the checks occurring. Clarification would be welcome!

On a related topic, we might consider setting the parent of the `QTimer` to be the `NapariQtNotification`, so destroying the latter also destroys the former. But maybe that also has similar underlying issues?

Reply:
Ah yeah, I forgot that we can connect to functors, so yeah I expect we're using that variant of the [`QObject.connect` method](https://doc.qt.io/qt-5/qobject.html#connect-4), in which case it does come with the following warning:

> The connection will automatically disconnect if the sender is destroyed. However, you should take care that any objects used within the functor are still alive when the signal is emitted.

The sender here is `QTimer` which can outlive the `NapariQtNotification` Python object used in that functor depending on the timing of Python's garbage collector. I'm unsure of a lot of the details here (e.g. how pyqt and pyside wrap Qt), but I see how that could cause problems even if there's no other thread.